crates/store/src/state.rs

```diff
@@ -332,6 +316,13 @@ impl State {
             // successfully.
             let mut inner = self.inner.write().await;
 
+            // We need to check that the account tree root hasn't changed while we were waiting for
+            // the DB update task to complete. If it has changed, we mustn't proceed with in-memory
+            // updates, since it might lead to inconsistent state.
+            if inner.account_tree.root() != account_tree_old_root {
+                return Err(ApplyBlockError::ConcurrentWrite);
+            }
```

Comment:
Why do we need this check here? If the account tree has changed somehow, wouldn't we get an error on line 344 below? Or more specifically: which scenario are we trying to prevent with this?

If we do need this check, why are we checking only the account tree and not nullifier tree as well?

Reply:
The motivation for adding this check was in the possible situation, when nullifier tree was successfully updated, but update of accounts tree raised an error, this could lead to inconsistent in-memory state. Additional check before updating of nullifier tree can prevent this.

Reply:
Makes sense. Could we move this check closer to the update site (e.g., to line 336 below) - or is there a reason it needs to be here?

Also, I'd expand the comment to explain the motivation more (like you've done in the comment). For example, there seems to be a specific order in which we need to do updates: first nullifier tree, then account tree, and lastly MMR - so, we should add a comment that this order is meaningful and shouldn't be changed.

Reply:
> Could we move this check closer to the update site (e.g., to line 336 below) - or is there a reason it needs to be here?

In this case we would allow database to update, but fail to update in-memory state. This is also not a good choice. By the way, we should add checking for nullifier tree root here as well in order to prevent this inconsistent between in-memory and db state. I will add it now.

Reply:
Done!